applications/benchmark/ShapeFunctionMemoization.cpp

```diff
@@ -4,61 +4,202 @@
 #include "mechanics/interpolationtypes/Interpolation3DTetrahedron.h"
 #include "mechanics/nodes/NodeEnum.h"
 #include "mechanics/interpolationtypes/InterpolationTypeEnum.h"
+#include "mechanics/elements/ElementShapeFunctions.h"
 
-#include "math/NaturalCoordinateMemoizer.h"
+#include <vector>
+#include <unordered_map>
 
-using Interpolation = NuTo::Interpolation3DTetrahedron;
-using Integration = NuTo::IntegrationType3D4NGauss4Ip;
-auto dof = NuTo::Node::eDof::DISPLACEMENTS;
-
-BENCHMARK(Hash, WithoutMemoization, runner)
+//! @brief transforms a vector of natural node coordinates to a reasonable id
+template <int TRaster>
+struct NaturalCoordianteToId
 {
-    Integration integration;
-    Interpolation interpolation(dof, NuTo::Interpolation::eTypeOrder::EQUIDISTANT2, 3);
-
-    while (runner.KeepRunningTime(1))
+    //! @brief 'rasterize' the natural coordinates (always [-1, 1]) into MaxId() equidistant blocks
+    template <typename TNaturalCoords>
+    size_t operator()(const TNaturalCoords& v) const
     {
-        for (int i = 0; i < integration.GetNumIntegrationPoints(); ++i)
+        int id = 0;
+        constexpr float eps = 1.e-2;
+        constexpr float factor = 0.5 * TRaster - 2 * eps;
+        // reverse the loop to get
+        // id = x + 10*y + 100*z
+        // instead of (unreversed)
+        // id = 100*x + 10*y + z
+        // IMO the first version is better to keep the data in 1D/2D closer together
+        for (int i = v.size() - 1; i >= 0; --i)
         {
-            auto coords = integration.GetLocalIntegrationPointCoordinates(i);
-            auto N = interpolation.CalculateMatrixN(coords);
+            assert(v[i] > -1. - eps);
+            assert(v[i] < 1. + eps);
+            id *= TRaster;
+            id += factor * (v[i] + 1 - eps); // transformation from float[-1, 1] --> int[0, TRaster]
         }
+        return id;
     }
-}
 
-//! @brief current NuTo implementation
-BENCHMARK(Hash, IdVectorHash, runner)
+    static constexpr size_t MaxId()
+    {
+        return TRaster * TRaster * TRaster;
+    }
+};
+
+
+//! @brief provides a memoization of the a std::function<TResult(TNaturalCoords)>, mainly used for
+//!        element shape functions and their derivatives
+//! @param TResult ... return type of the function
+//! @param TNaturalCoords ... vector type of natural coordinates
+//! @param TIdHash ... function object that transforms TNaturalCoords into a vector id
+template <typename TResult, typename TNaturalCoords, typename TIdHash = NaturalCoordianteToId<32>>
```

Comment:
This benchmark contains a more efficient (but unsafer, more error prone) way of the memoization using `std::vector` instead of `std::map`. And indeed, the choice of TRaster is important. It has to be big enough to eliminate __clashes__ in one block - so that the different natural coordinates are not assigned to the same ID. It is easily possible to _detect_ these clashes with some additional lines of code and throw an exception. But how would the user solve the problem? He has to look at the implementation... So I went for the slightly more inefficient implementation as it is now. The code in this benchmark could be the starting point for a more efficient implementation. The goal here should be to actually __solve__ the ID clash instead of just detecting it. I am convinced that this is possible.